llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp

```diff
@@ -7428,12 +7428,98 @@ bool CombinerHelper::matchNonNegZext(const MachineOperand &MO,
   LLT SrcTy = MRI.getType(Src);
   const auto &TLI = getTargetLowering();
 
+  const MachineFunction &MF = *MO.getParent()->getMF();
+  LLVMContext &Ctx = MF.getFunction().getContext();
+  auto &DL = MF.getDataLayout();
+
   // Convert zext nneg to sext if sext is the preferred form for the target.
   if (isLegalOrBeforeLegalizer({TargetOpcode::G_SEXT, {DstTy, SrcTy}}) &&
-      TLI.isSExtCheaperThanZExt(getMVTForLLT(SrcTy), getMVTForLLT(DstTy))) {
+      TLI.isSExtCheaperThanZExt(SrcTy, DstTy, DL, Ctx)) {
     MatchInfo = [=](MachineIRBuilder &B) { B.buildSExt(Dst, Src); };
     return true;
   }
 
   return false;
 }
+
+bool CombinerHelper::matchZextInteger(const MachineInstr &MI,
+                                      APInt &MatchInfo) {
+  const GZext *Zext = cast<GZext>(&MI);
+
+  std::optional<APInt> Input = getIConstantVRegVal(Zext->getSrcReg(), MRI);
+  if (!Input)
+    return false;
+
+  LLT DstTy = MRI.getType(Zext->getReg(0));
+
+  if (!isConstantLegalOrBeforeLegalizer(DstTy))
+    return false;
+
+  MatchInfo = Input->zext(DstTy.getScalarSizeInBits());
+
+  return true;
+}
+
+bool CombinerHelper::matchSextInteger(const MachineInstr &MI,
+                                      APInt &MatchInfo) {
+  const GSext *Sext = cast<GSext>(&MI);
+
+  std::optional<APInt> Input = getIConstantVRegVal(Sext->getSrcReg(), MRI);
+  if (!Input)
+    return false;
+
+  LLT DstTy = MRI.getType(Sext->getReg(0));
+
+  if (!isConstantLegalOrBeforeLegalizer(DstTy))
+    return false;
+
+  MatchInfo = Input->sext(DstTy.getScalarSizeInBits());
+
+  return true;
+}
+
+bool CombinerHelper::matchTruncInteger(const MachineInstr &MI,
+                                       APInt &MatchInfo) {
+  const GTrunc *Trunc = cast<GTrunc>(&MI);
+
+  std::optional<APInt> Input = getIConstantVRegVal(Trunc->getSrcReg(), MRI);
+  if (!Input)
+    return false;
+
+  LLT DstTy = MRI.getType(Trunc->getReg(0));
+
+  if (!isConstantLegalOrBeforeLegalizer(DstTy))
+    return false;
+
+  MatchInfo = Input->trunc(DstTy.getScalarSizeInBits());
+
+  return true;
+}
```

Comment:
The three functions above have effectively the same definitions. Is it possible to merge them?

Reply:
They differ in : zext, sext, and trunc on APInt. It is hard to merge.

Reply:
I still think it would be better if they were merged, with the `cast<>` operand and the `Input->` member functions being template parameters to a generic function.